diff --git a/althaia/marshmallow/fields.py b/althaia/marshmallow/fields.py
index 46e3744..36b2a2e 100644
--- a/althaia/marshmallow/fields.py
+++ b/althaia/marshmallow/fields.py
@@ -1302,8 +1346,11 @@ class _TemporalField(Field[_D], metaclass=abc.ABCMeta):
                 "invalid", input=value, obj_type=self.OBJ_TYPE
             ) from error
 
-    @staticmethod
-    @abc.abstractmethod
+    # this is deprecated, but the original (staticmethod + abstractmethod) fails to compile in cython
+    # references:
+    # https://github.com/cython/cython/issues/2651
+    # https://github.com/cython/cython/issues/1434
+    @abc.abstractstaticmethod
     def _make_object_from_format(value: typing.Any, data_format: str) -> _D: ...
 
 
